### ISSUE TITLE
[Bug] MLA KV writers can overwrite reserved padding slot 0

### ISSUE THREAD

**Issue by tanth47** [sgl-project/sglang#36207]
## Summary

SGLang reserves physical KV-cache slot 0 as a finite padding source for CUDA
graph and DP-attention padding. Generic KV stores protect this invariant after
#32477, but MLA-specific writers can still accept `loc == 0` and overwrite the
reserved slot with undefined padding rows.

If a padding row contains NaNs, the write poisons slot 0. A later padded or
masked attention row can then read that slot and propagate NaNs into attention
output and model logits, producing silently corrupted text.

## Environment where the failure was isolated

- 8 x AMD MI350X (`gfx950`)
- ROCm 7.2.4
- Official image: `lmsysorg/sglang-rocm:v0.5.18-rocm724-mi35x-20260822`
- SGLang revision: `db570fe619ae15d8bd4ee04f146bc8234c52a20d`
- Model: `amd/GLM-5.2-MXFP4`, real weights
- KV cache: BF16
- TP8, DP4 with DP attention, EP8, MTP
- DSA attention with TileLang prefill/decode

## User-visible symptom

During the investigation, affected concurrent requests collapsed into very
short nonsensical text followed by one repeated token. One 64-token response
was:

```text
(... thess,
```

with output IDs ending in repeated token `154822`:

```text
[320, 1112, 279, 778, 11, 154822, 154822, 154822, ...]
```

The unpatched real-weight runtime-reference gate reported 438 bad comparison
events and 4,914 bad TileLang rows. After protecting the producer-side reserved
slot, the same c8 x 4 class of gate completed with zero tensor anomalies and
zero strong output failures.

## Deterministic writer reproduction

The failure does not require relying on allocator garbage. Initialize slot 0 to
a finite value, use `loc=[0, 2]`, and make the source row mapped to slot 0 NaN:

```python
kv = torch.full((4, 1, 576), 7.0, dtype=torch.bfloat16, device="cuda")
nope = torch.stack(
    [
        torch.full((1, 512), float("nan"), dtype=torch.bfloat16, device="cuda"),
        torch.full((1, 512), 17.0, dtype=torch.bfloat16, device="cuda"),
    ]
)
rope = torch.stack(
    [
        torch.full((1, 64), float("nan"), dtype=torch.bfloat16, device="cuda"),
        torch.full((1, 64), 19.0, dtype=torch.bfloat16, device="cuda"),
    ]
)
loc = torch.tensor([0, 2], dtype=torch.int64, device="cuda")

slot0_before = kv[0].clone()
set_mla_kv_buffer_triton(kv, loc, nope, rope)
torch.cuda.synchronize()

assert torch.equal(kv[0], slot0_before)  # Fails without the reserved-slot guard
assert torch.isfinite(kv[2]).all()       # The valid positive slot is still written
```

The focused regression in #36003 covers BF16, FP8 quantized, scale-buffer, and
CUDA TMA/JIT writers, including int32/int64 locations and the explicit opt-out
path.

The production GLM ROCm fused writer has an even smaller deterministic probe:
with `slot_mapping=[0]` and NaN padding K, unpatched
`aiter.fused_qk_rope_cat_and_cache_mla` changes slot 0 from finite/all-zero to
576 NaNs. With the downstream fused-writer guard, slot 0 remains finite and
unchanged.

## Root cause

The generic store path and MLA paths currently disagree on the meaning of
physical slot 0:

| Writer path | Current behavior |
| --- | --- |
| Generic `store_cache()` after #32477 | Skips the configured reserved index 0 |
| SGLang MLA BF16/FP8/scale writers | Treat `loc == 0` as a valid destination |
| SGLang CUDA TMA/JIT MLA writer | Treats `loc == 0` as a valid destination |
| AITER fused GLM MLA rope/cat/cache writer | Treats `slot_mapping == 0` as valid and bypasses the SGLang writers |

On DCP rank 0 and in graph/DP-padding rows, `0` can appear as a padding
location. The source values for those rows are not required to be initialized,
so accepting the destination mutates the supposedly finite padding slot.

## Fix scope

PR #36003 extends the #32477 contract to SGLang-owned MLA writers:

- add `reserved_skip_index=0` to BF16, FP8-quantized, and scale writers;
- apply the same guard to the CUDA TMA/JIT writer;
- retain `reserved_skip_index=-1` as an explicit opt-out;
- preserve launch shapes and valid positive-slot behavior.

The fused AITER GLM path writes KV internally and invokes attention with
`save_kv_cache=False`, so it bypasses PR #36003. That integration needs a
separate API-safe follow-up; this issue should not be considered fully resolved
for GLM ROCm by #36003 alone.

## Requested outcome

1. Merge #36003 to make all SGLang-owned MLA writers obey the reserved-slot
   invariant.
2. Agree on the fused AITER integration contract without globally assuming
   that slot 0 is invalid for every AITER caller.
3. Keep a deterministic poison-row regression for every writer that can receive
   graph or DP-padding locations.

**Comment by tanth47** [comment]
Upstream fix proposed in ROCm/aiter#5010. It adds a backward-compatible `pad_slot_id` opt-in to the fused MLA writer and covers both KV-write branches with regression tests. If it lands, the SGLang follow-up only needs to pass `pad_slot_id=0` on this path.
